File: src/limiter/TokenBucket.js
import BaseLimiter from "./BaseLimiter.js";
import { loadLuaScript } from "../lua/loadLuaScript.js";

class TokenBucket extends BaseLimiter {

    constructor({ redis, capacity, refillRate, interval }) {
        super({ redis });

        this.validatePositiveInteger(capacity, "capacity");
        this.validatePositiveInteger(refillRate, "refillRate");
        this.validatePositiveInteger(interval, "interval");

        this.capacity = capacity;
        this.refillRate = refillRate;
        this.interval = interval;

        this.scriptSha = null;
    }

    async getScriptSha() {
        if (!this.scriptSha) {
            this.scriptSha = await loadLuaScript(
                this.redis,
                "tokenBucket.lua"
            );
        }

        return this.scriptSha;
    }

    async consume(key) {

        const tokenKey = `${key}:tokens`;
        const timeKey = `${key}:lastRefill`;
        const now = Date.now();

        let sha = await this.getScriptSha();

        try {
            const [allowed, remaining, retryAfter] = await this.redis.evalSha(
                sha,
                {
                    keys: [tokenKey, timeKey],
                    arguments: [
                        this.capacity.toString(),
                        this.refillRate.toString(),
                        this.interval.toString(),
                        now.toString()
                    ]
                }
            );

            if (allowed === 1) {
                return this.createResponse({
                    allowed: true,
                    limit: this.capacity,
                    remaining
                });
            }

            return this.createResponse({
                allowed: false,
                limit: this.capacity,
                remaining: 0,
                retryAfter
            });

        } catch (err) {
            if (err.message.includes("NOSCRIPT")) {
                this.scriptSha = null;

                sha = await this.getScriptSha();

                const [allowed, remaining, retryAfter] = await this.redis.evalSha(
                    sha,
                    {
                        keys: [tokenKey, timeKey],
                        arguments: [
                            this.capacity.toString(),
                            this.refillRate.toString(),
                            this.interval.toString(),
                            now.toString()
                        ]
                    }
                );

                if (allowed === 1) {
                    return this.createResponse({
                        allowed: true,
                        limit: this.capacity,
                        remaining
                    });
                }

                return this.createResponse({
                    allowed: false,
                    limit: this.capacity,
                    remaining: 0,
                    retryAfter
                });
            }

            throw err;
        }
    }
}

export default TokenBucket;

File: src/lua/tokenBucket.lua
local tokenKey = KEYS[1]
local timeKey = KEYS[2]

local capacity = tonumber(ARGV[1])
local refillRate = tonumber(ARGV[2])
local interval = tonumber(ARGV[3])
local now = tonumber(ARGV[4])

local storedTokens = redis.call("GET", tokenKey)
local storedLastRefill = redis.call("GET", timeKey)

local tokens
local lastRefill

if not storedTokens or not storedLastRefill then
    tokens = capacity
    lastRefill = now
else
    tokens = tonumber(storedTokens)
    lastRefill = tonumber(storedLastRefill)
end

local elapsed = (now - lastRefill) / 1000
local intervalsPassed = math.floor(elapsed / interval)

if intervalsPassed > 0 then
    tokens = math.min(
        capacity,
        tokens + intervalsPassed * refillRate
    )

    lastRefill = lastRefill + intervalsPassed * interval * 1000
end

if tokens <= 0 then

    local retryAfter = math.ceil(
        interval - (elapsed % interval)
    )

    redis.call("SET", tokenKey, tokens)
    redis.call("SET", timeKey, lastRefill)

    return {
        0,
        tokens,
        retryAfter
    }
end

tokens = tokens - 1

redis.call("SET", tokenKey, tokens)
redis.call("SET", timeKey, lastRefill)

return {
    1,
    tokens,
    0
}